### PR TITLE
Add CoreV3 Engine fork for Prohibition

### PR DESCRIPTION
**Pull request by tlip**
## Description of the change

Fork of the CoreV3 Engine Flex contract (and some supporting contracts) to add:
* The ability to have artists create their own project shells
* The ability for "verified" or trusted artists activate their own projects
* Ability for "child" admin accounts that are given rights to call specific selectors

> reminder: Any mainnet deployments after 10 Jan 2023 should be tagged as [Releases](https://github.com/ArtBlocks/artblocks-contracts/releases) in this repository. This ensures that the deployed contract source code is easily verifiable by anyone.

### CODE REVIEW

File: contracts/engine/V3/forks/PROHIBITION/AdminACLV0_PROHIBITION.sol [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+ * @author Art Blocks Inc.
+ * @notice Privileged Roles and Ownership:
+ * This contract has a single superAdmin that passes all ACL checks. All checks
+ * for any other address will return false.
+ * The superAdmin can be changed by the current superAdmin.
+ * Care must be taken to ensure that the admin ACL contract is secure behind a
+ * multi-sig or other secure access control mechanism.
+ */
+contract AdminACLV0_PROHIBITION is IAdminACLV0_PROHIBITION, ERC165 {
+    string public AdminACLType = "AdminACLV0";
```
Comment (jakerockland) [inline contracts/engine/V3/forks/PROHIBITION/AdminACLV0_PROHIBITION.sol:22]: Suggestion to please make this a custom named type (e.g. `"AdminACLV0_PROHIBITION"`)

File: contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol [hunk 2]
```diff
@@ -0,0 +1,2276 @@
...
+import "../../../../interfaces/0.8.x/IDependencyRegistryCompatibleV0.sol";
+import "../../../../interfaces/0.8.x/IManifold.sol";
+
+import "@openzeppelin-4.7/contracts/access/Ownable.sol";
+import "../../../../libs/0.8.x/ERC721_PackedHashSeed.sol";
+import "../../../../libs/0.8.x/BytecodeStorage.sol";
+import "../../../../libs/0.8.x/Bytes32Strings.sol";
+
+/**
+ * @title Art Blocks Engine Flex ERC-721 core contract, V3.
```
Comment (jakerockland) [inline contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol:20]: Nit: A general suggestion to add a `V3_PROHIBTION_CHANGELOG.md` file to this `engine/V3/forks/PROHIBITION` directory that summarizes the small diffs here for future reference, while it is fresh-of-mind.

**Review by jakerockland — APPROVED** [review]
Some minor suggested edits here, but generally speaking overall LGTM.

@ryley-o would you be willing to TAL this week as well for a second set of eyes from our end.

Admin ACL diff: https://www.diffchecker.com/o9oOYbzW/
CoreV3 Flex diff: https://www.diffchecker.com/yYr25IDS/
MinterFilter diff: https://www.diffchecker.com/Pphlwqvi/

File: contracts/engine/V3/forks/PROHIBITION/AdminACLV0_PROHIBITION.sol (hunk 1 above)
Comment (tlip) [reply contracts/engine/V3/forks/PROHIBITION/AdminACLV0_PROHIBITION.sol:22]: Ah! I meant to do that too. Thanks. for catching this.

Should I update the `coreType` in the corev3 contract as well?

File: contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol (hunk 2 above)
Comment (tlip) [reply contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol:20]: Done.

File: contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol [hunk 3]
```diff
@@ -0,0 +1,2276 @@
...
+     * to call function with selector `_selector` on contract `_contract`, as
+     * determined by this contract's current Admin ACL contract. Expected use
+     * cases include minter contracts checking if caller is allowed to call
+     * admin-gated functions on minter contracts.
+     * @param _sender Address of the sender calling function with selector
+     * `_selector` on contract `_contract`.
+     * @param _contract Address of the contract being called by `_sender`.
+     * @param _selector Function selector of the function being called by
+     * `_sender`.
+     * @return bool Whether `_sender` is allowed to call function with selector
```
Comment (ryley-o) [inline contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol:2042]: nit: add natspec for param _projectId

File: contracts/engine/V3/forks/PROHIBITION/AdminACLV0_PROHIBITION.sol (hunk 1 above)
Comment (jakerockland) [reply contracts/engine/V3/forks/PROHIBITION/AdminACLV0_PROHIBITION.sol]: > Should I update the coreType in the corev3 contract as well?

I don't believe so, though will defer to @ryley-o on this one.

File: contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol (hunk 2 above)
Comment (jakerockland) [reply contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol:20]: 🙏

**Review by ryley-o — CHANGES_REQUESTED** [review]
Thanks for developing this fork!
A couple requests:
- We will certainly want to keep the existing `adminACLAllowed` public function (without projectId) to maintain compatibility in our future state of a shared minter suite (including shared minter filter). Additionally, some current minters would need forking to be compatible today. Please incorporate the changes highlighted in the following diff (apologies for some formatting changes that are irrelevant but show up in the diff): https://www.diffchecker.com/9cDsWC13/
  - note: switched some constants to state variables to keep strings out of bytecode to stay <24kb
  - note: you will want to update the `IGenArt721CoreV3_Engine_PROHIBITION.sol` to include the function as well
- I think implementing the above changes will allow you to be able to not have to fork MinterFilterV1
- Appreciate the updated tests, but would appreciate if all new core test logic lived in subdirectories. I think this can be as simple as copying your new files to a subdirectory, and updating the `coreContractsToTest` to be just the PROHIBITION fork (while leaving the original non-forked test files unchanged for this PR)
  - The exception here is that totally good with the current edits to `util/common.ts`, which LGTM as-is to me
  - Also note that if you no longer have to fork the MinterFilter, I think your tests simplify a bit!

Thank you!

File: contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol [hunk 4]
```diff
@@ -0,0 +1,2276 @@
...
+    /// that determines whether or not admin approval^ should be required
+    /// to accept artist address change proposals, or if these proposals
+    /// should always auto-approve, as determined by the business process
+    /// requirements of the Engine partner using this contract.
+    ///
+    /// ^does not apply in the case where contract-ownership itself is revoked
+    bool public immutable autoApproveArtistSplitProposals;
+
+    /// version & type of this core contract
+    bytes32 constant CORE_VERSION = "v3.1.2";
```
Comment (ryley-o) [inline contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol:274]: please update this to be `"v3.1.3"`, thanks!

File: contracts/engine/V3/forks/PROHIBITION/AdminACLV0_PROHIBITION.sol (hunk 1 above)
Comment (ryley-o) [reply contracts/engine/V3/forks/PROHIBITION/AdminACLV0_PROHIBITION.sol]: leave `coreType` unchanged (`GenArt721CoreV3_Engine_Flex`) 👍 That will let our indexing system treat it as an Engine Flex contract, which is what we want in this case 😃

File: contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol (hunk 3 above)
Comment (tlip) [reply contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol:2043]: Not sure why this is still listed as `outdated`. Definitely updated.

**Review by ryley-o — CHANGES_REQUESTED** [review]
This looks great, and I have one more requested change 🙏 

Thinking on this more, and I am going to follow up with an update our Engine docs to better communicate this, I would prefer all Engine forks conform to our existing interfaces. We design our data indexing layers based off of the interfaces, so enforcing interface at the compiler level will greatly reduce the chance of any integration issues in the future (which is a win-win for everyone).

Based on that, would appreciate if you could implement the changes described in this commit, which essentially changes `IGenArt721CoreContractV3_Engine_Flex_PROHIBITION` to simply extend `IGenArt721CoreContractV3_Engine_Flex`, accomplishing all the guarantees we would like to achieve: https://github.com/ArtBlocks/artblocks-contracts/pull/644/commits/776b78ffe9df8e9538167dabff5eed3ee828d969#diff-c18369ac7400114ef07e9ad79b25b8bf2edcfe93654bc0bde3bc15c7b3dbc1db

Appreciate your patience on all of these reviews today! 🙏 😄

File: test/core/V3/GenArt721CoreV3_Integration.test.ts [hunk 5]
```diff
@@ -260,12 +260,11 @@ for (const coreContractName of coreContractsToTest) {
...
-          config.genArt721Core
-            .connect(config.accounts.deployer)
-            .addProject("new project", config.accounts.artist2.address),
-          "Only Admin ACL allowed"
-        );
+        const addProjectCall = config.genArt721Core
+          .connect(config.accounts.deployer)
+          .addProject("new project", config.accounts.artist2.address);
+
+        await expectRevert(addProjectCall, "Only Admin ACL allowed");
```
[the remark this answers is not in the document]
Comment (tlip) [reply test/core/V3/GenArt721CoreV3_Integration.test.ts:267]: Oh right. My bad.

File: contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol [hunk 6]
```diff
@@ -0,0 +1,2311 @@
...
+ * - toggleProjectIsActive
+ * - addProject
+ * - forbidNewProjects (forever forbidding new projects)
+ * - updateDefaultBaseURI (used to initialize new project base URIs)
+ * - updateIPFSGateway
+ * - updateArweaveGateway
+ * ----------------------------------------------------------------------------
+ * The following functions are restricted to either the the Artist address or
+ * the Admin ACL contract, only when the project is not locked:
+ * - toggleProjectIsPaused (note the artist can still mint while paused)
```
[the remark this answers is not in the document]
Comment (tlip) [reply contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol:50]: Ah. I did at one point. Changed it back though.

File: test/core/V3/GenArt721CoreV3_Integration.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (tlip) [reply test/core/V3/GenArt721CoreV3_Integration.test.ts:267]: My bad. Thought I cleared everything out.

File: contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (tlip) [reply contracts/engine/V3/forks/PROHIBITION/GenArt721CoreV3_Engine_Flex_PROHIBITION.sol]: My original changes were pretty massive :/. Must've forgotten to change that one back.

**Review by ryley-o — APPROVED** [review]
LGTM 🎉 🙏 🙏

File: test/core/V3/prohibition/GenArt721CoreV3_Integration_PROHIBITION.test.ts [hunk 7]
```diff
@@ -0,0 +1,722 @@
...
+    describe("coreVersion", function () {
+      it("returns expected value", async function () {
+        const config = await loadFixture(_beforeEach);
+        let targetCoreVersion;
+        if (coreContractName === "GenArt721CoreV3") {
+          targetCoreVersion = "v3.0.2";
+        } else if (coreContractName === "GenArt721CoreV3_Explorations") {
+          targetCoreVersion = "v3.0.3";
+        } else if (coreContractName.includes("GenArt721CoreV3_Engine")) {
+          targetCoreVersion = "v3.1.2";
```
Comment (ryley-o) [inline test/core/V3/prohibition/GenArt721CoreV3_Integration_PROHIBITION.test.ts:360]: this test is failing - can you update to handle the prohibition forked case (`v3.1.3`), thanks!

**Comment by ryley-o** [comment]
@tlip do you mind merging in the latest changes to main to this branch? thanks!

File: test/core/V3/prohibition/GenArt721CoreV3_Integration_PROHIBITION.test.ts (hunk 7 above)
Comment (tlip) [reply test/core/V3/prohibition/GenArt721CoreV3_Integration_PROHIBITION.test.ts:360]: I must be losing my mind. I could've sworn I did this... or was it somewhere else?

Comment (tlip) [reply test/core/V3/prohibition/GenArt721CoreV3_Integration_PROHIBITION.test.ts]: done

Comment (ryley-o) [reply test/core/V3/prohibition/GenArt721CoreV3_Integration_PROHIBITION.test.ts]: i think this test is redundant with one in the views tests, so I promise you aren't losing your mind 😅

**Pull request merged**